### PR TITLE
all: add support for v4 OpenAPI

### PR DESCRIPTION
**Pull request by jonathaningram**
This is done in a progressive way to allow callers to opt in to using
the v4 API.

### CODE REVIEW

File: opts.go [hunk 1]
```diff
@@ -13,8 +32,16 @@ func WithEndpoint(endpoint string) Opt {
...
-	return WithGRPCDialOption(grpc.WithPerRPCCredentials(creds{c}))
+func WithCredentials(cr Credentials) Opt {
+	return func(c *Client) {
+		if kc, ok := cr.(keyCreds); ok {
+			c.openAPI.auth = openapi.BasicAuth{
+				UserName: kc.keyID,
+				Password: kc.keySecret,
+			}
+		}
+		WithGRPCDialOption(grpc.WithPerRPCCredentials(creds{cr}))(c)
```
Comment (benhinchley) [inline opts.go]: You will need to return the `WithGRPCDialOption`
```suggestion
		return WithGRPCDialOption(grpc.WithPerRPCCredentials(creds{cr}))(c)
```

**Review by benhinchley — COMMENTED** [review]
I would just update this to use the v4 client and drop the gRPC stuff.

**Comment by jonathaningram** [comment]
> I would just update this to use the v4 client and drop the gRPC stuff.

The intention was that we could progressively switch to using v4 rather than having all existing code frozen on the HEAD version of sdk-go until they upgraded.

**Comment by benhinchley** [comment]
> > I would just update this to use the v4 client and drop the gRPC stuff.
> 
> The intention was that we could progressively switch to using v4 rather than having all existing code frozen on the HEAD version of sdk-go until they upgraded.

Yeah that makes sense 👍

File: opts.go (hunk 1 above)
Comment (benhinchley) [reply opts.go]: Oh sorry saw just saw the little `(c)` at the end. hmm I would maybe make that a bit more obvious and break

```
gRPCCredentialsOpt := WithGRPCDialOption(grpc.WithPerRPCCredentials(creds{cr}))
gRPCCredentialsOpt(c)
```

Comment (jonathaningram) [reply opts.go]: @benhinchley done. Thanks.

File: record.go [hunk 2]
```diff
@@ -197,6 +198,34 @@ func (c *Client) DeleteRecord(ctx context.Context, k *Key) error {
...
+					case codes.NotFound:
+						return nil, fmt.Errorf("%v: %w", k, ErrNoSuchRecord)
+					}
+				}
+			}
+			return nil, fmt.Errorf("could not get record: %w", err)
+		}
+
+		return resp, nil
+	}
```
Comment (benhinchley) [inline record.go]: I would maybe break this out into another method so this looks like
```
if c.v4 {
    return c.getRecordV4(ctx, k)
}

// ... rest of code
}
```

Comment (jonathaningram) [reply record.go]: @benhinchley done

File: record.go [hunk 3]
```diff
@@ -192,11 +193,43 @@ func (c *Client) DeleteRecord(ctx context.Context, k *Key) error {
...
+			Value: fmt.Sprintf("%v", k.value),
+		},
+	}
+	resp, _, err := c.openAPI.client.RecordsApi.GetRecord(ctx, c.Collection).Sajariv4beta1GetRecordRequest(req).Execute()
+	if err != nil {
+		switch x := err.(type) {
+		case openapi.GenericOpenAPIError:
+			m := x.Model()
+
+			if m, ok := m.(openapi.GatewayruntimeError3); ok {
```
Comment (benhinchley) [inline record.go:211]: Could this be? 
```suggestion
			if m, ok := m.(interface { GetCode() int }); ok {
```

**Pull request merged**